### PR TITLE
memstats collection

### PR DESCRIPTION
**Pull request by KevinPike**
This is still a work in progress. 

Is this along the lines for what you had in mind for #96 @beorn7? I'm new to prometheus and this client library, feedback is greatly appreciated.

### CODE REVIEW

File: prometheus/go_collector.go [hunk 1]
```diff
@@ -23,6 +26,7 @@ func NewGoCollector() *goCollector {
 			"go_gc_duration_seconds",
 			"A summary of the GC invocation durations.",
 			nil, nil),
+		memStatsCollector: newMemStatsCollector(),
```
Comment (beorn7) [inline prometheus/go_collector.go]: If you really want to wrap the memstats in a different collector, then you have to call its Describe and Collect methods in the goCollector's Describe and Collect methods.

File: prometheus/go_collector.go [hunk 2]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+// newMemStatsCollector creates a new runtime.MemStats collector
+func newMemStatsCollector() *memStatsCollector {
+	return &MemStatsCollector{
+		alloc: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "alloc_bytes",
+			Help:      "bytes allocated and still in use",
+		}),
+		totalAlloc: prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (beorn7) [inline prometheus/go_collector.go]: This is a counter.

File: prometheus/go_collector.go [hunk 3]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+}
+
+// newMemStatsCollector creates a new runtime.MemStats collector
+func newMemStatsCollector() *memStatsCollector {
+	return &MemStatsCollector{
+		alloc: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "alloc_bytes",
+			Help:      "bytes allocated and still in use",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Please capitalize and punctuate help string as English sentences.

File: prometheus/go_collector.go [hunk 4]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+		totalAlloc: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "alloc_bytes_total",
+			Help:      "bytes allocated (even if freed)",
+		}),
+		sys: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "sys",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Add unit.

File: prometheus/go_collector.go [hunk 5]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+			Name:      "alloc_bytes",
+			Help:      "bytes allocated and still in use",
+		}),
+		totalAlloc: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "alloc_bytes_total",
+			Help:      "bytes allocated (even if freed)",
+		}),
+		sys: prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (beorn7) [inline prometheus/go_collector.go]: Is this a counter? (If yes, prepend `_total` to the name.)

File: prometheus/go_collector.go [hunk 6]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "sys",
+			Help:      "bytes obtained from system",
+		}),
+		lookups: prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (beorn7) [inline prometheus/go_collector.go]: Is this a counter?  (If yes, prepend `_total` to the name.)

File: prometheus/go_collector.go [hunk 7]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "lookups",
+			Help:      "number of pointer lookups",
+		}),
+		mallocs: prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (beorn7) [inline prometheus/go_collector.go]: Is this a counter?  (If yes, prepend `_total` to the name.)

File: prometheus/go_collector.go [hunk 8]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "mallocs",
+			Help:      "number of mallocs",
+		}),
+		frees: prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (beorn7) [inline prometheus/go_collector.go]: Is this a counter?  (If yes, prepend `_total` to the name.)

File: prometheus/go_collector.go [hunk 9]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "frees",
+			Help:      "number of frees",
+		}),
+		heapAlloc: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "go",
+			Subsystem: "memstats",
+			Name:      "heap_alloc",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Prepend `_bytes`.

File: prometheus/go_collector.go [hunk 10]
```diff
@@ -48,3 +52,153 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
+			Help:      "bytes allocated and still in use",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Add "heap" somewhere.

**Comment by beorn7** [comment]
Thanks. See comments to get the spirit of metric naming, typing, and help strings etc.

I'd probably not create a separate memStatsCollector, but perhaps a struct of Descs that mirrors MemStats. Then collect metrics on the fly as `MustNewConstMetric` in the Collect method (similar to the GC stats, and the goroutine gauge should actually do the same).

**Comment by beorn7** [comment]
Or perhaps even put all the Descs in a dict, which makes Describe() way easier to implement.

File: prometheus/go_collector.go (hunk 1 above)
Comment (KevinPike) [inline prometheus/go_collector.go]: I'm not set on having it in a different collector. I will un-embed it.

**Comment by KevinPike** [comment]
Thanks for the great feedback! I removed the separate memStatsCollector, used counters where you suggested, and improved names and descriptions. I'm going to look into `MustNewConstMetric`.

**Comment by beorn7** [comment]
@KevinPike Sorry, this seems to have fallen through the cracks. I got the impression you were still working on it before you wanted another pass of review.

If you feel this is ready for review, I'm happy to have a look. Just let me know.

**Comment by KevinPike** [comment]
@beorn7 no problem. I just addressed the last comment which was to not have a separate memstats collector. It is ready for review now :smile:

File: prometheus/go_collector.go [hunk 11]
```diff
@@ -9,27 +10,245 @@ import (
...
+	ms *runtime.MemStats
+	// metrics to describe and collect
+	metrics memStatsMetrics
 }
 
 // NewGoCollector returns a collector which exports metrics about the current
 // go process.
 func NewGoCollector() *goCollector {
 	return &goCollector{
+		ms: new(runtime.MemStats),
```
Comment (beorn7) [inline prometheus/go_collector.go]: new(runtime.MemStats) -> &runtime.MemStats{}
See styleguide as linked from https://github.com/prometheus/client_golang/blob/master/CONTRIBUTING.md .

File: prometheus/go_collector.go [hunk 12]
```diff
@@ -47,4 +266,16 @@ func (c *goCollector) Collect(ch chan<- Metric) {
...
 	quantiles[0.0] = stats.PauseQuantiles[0].Seconds()
 	ch <- MustNewConstSummary(c.gcDesc, uint64(stats.NumGC), float64(stats.PauseTotal.Seconds()), quantiles)
+
+	runtime.ReadMemStats(c.ms)
+	for _, i := range c.metrics {
+		ch <- MustNewConstMetric(i.desc, i.valType, i.eval(c.ms))
+	}
+}
+
+// memStatsMetrics provide description, value, and value type for memstat metrics
```
Comment (beorn7) [inline prometheus/go_collector.go]: Period at the end.

File: prometheus/go_collector.go [hunk 13]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("alloc_bytes_total"),
+					"Total number of bytes allocated, even if freed.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.TotalAlloc) },
+				valType: CounterValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("sys_bytes"),
+					"Number of bytes obtained from system",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Period missing.

File: prometheus/go_collector.go [hunk 14]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("frees_total"),
+					"Total number of frees.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.Frees) },
+				valType: CounterValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("heap_alloc_bytes"),
+					"Number heap bytes allocated and still in use.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: "of" missing after "Number".

File: prometheus/go_collector.go [hunk 15]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("heap_sys_bytes"),
+					"Total bytes in heap obtained from system.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("heap_idle_bytes"),
+					"Number bytes in heap waiting to be used.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: "of" missing.

File: prometheus/go_collector.go [hunk 16]
```diff
@@ -9,27 +10,245 @@ import (
...
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapInuse) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("heap_released_bytes"),
+					"Number of bytes in heap released to OS.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapReleased) },
+				valType: GaugeValue,
```
Comment (beorn7) [inline prometheus/go_collector.go]: That is probably a counter. What do you think?

File: prometheus/go_collector.go [hunk 17]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("stack_bytes_inuse"),
+					"Number of bytes in use by the stack allocator.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.StackInuse) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("stack_sys_bytes"),
+					"Number of bytes in obtained from system for stack allocator.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Spurious "in".

File: prometheus/go_collector.go [hunk 18]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("mcache_sys"),
+					"Number of mcache structures obtained from system.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.MCacheSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("buck_hash_sys"),
+					"Profiling bucket hash table.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: I guess this should be "Number of bytes used by the profiling bucket hash table."

File: prometheus/go_collector.go [hunk 19]
```diff
@@ -9,27 +10,245 @@ import (
...
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.BuckHashSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("gc_metadata"),
+					"GC metadata.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: I guess this should also be "Number of bytes used for..."

File: prometheus/go_collector.go [hunk 20]
```diff
@@ -9,27 +10,245 @@ import (
...
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.GCSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("other_sys"),
+					"Other system allocations.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: See above.

File: prometheus/go_collector.go [hunk 21]
```diff
@@ -9,27 +10,245 @@ import (
...
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.OtherSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("next_gc"),
+					"Next collection will happen when HeapAlloc ≥ this amount.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Use the metric name, not HeapAlloc.

File: prometheus/go_collector.go [hunk 22]
```diff
@@ -9,27 +10,245 @@ import (
...
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.NextGC) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("last_gc"),
+					"End time of last garbage collection (nanoseconds since 1970).",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Please convert to seconds.

File: prometheus/go_collector.go [hunk 23]
```diff
@@ -9,27 +10,245 @@ import (
...
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.LastGC) },
+				valType: CounterValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("pause_total"),
+					"Total garbage collection pauses for all collections.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: in seconds. Please convert below.

File: prometheus/go_collector.go [hunk 24]
```diff
@@ -9,27 +10,245 @@ import (
...
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.NextGC) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("last_gc"),
+					"End time of last garbage collection (nanoseconds since 1970).",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.LastGC) },
+				valType: CounterValue,
```
Comment (beorn7) [inline prometheus/go_collector.go]: It's a timestamp, so use GaugeValue.

File: prometheus/go_collector.go (hunk 23 above)
Comment (beorn7) [inline prometheus/go_collector.go]: Actually, the total pause time is already contained in the GC summary. It should not be repeated here.

File: prometheus/go_collector.go [hunk 25]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("pause_total"),
+					"Total garbage collection pauses for all collections.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.PauseTotalNs) },
+				valType: CounterValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("gc_total"),
+					"Number of garbage collection.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Already contained in the GC summary. Please remove here.

File: prometheus/go_collector.go [hunk 26]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("next_gc"),
+					"Next collection will happen when HeapAlloc ≥ this amount.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.NextGC) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("last_gc"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: last_gc_time_seconds

File: prometheus/go_collector.go [hunk 27]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("other_sys"),
+					"Other system allocations.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.OtherSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("next_gc"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: next_gc_bytes

File: prometheus/go_collector.go [hunk 28]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("gc_metadata"),
+					"GC metadata.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.GCSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("other_sys"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: other_sys_bytes

File: prometheus/go_collector.go [hunk 29]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("buck_hash_sys"),
+					"Profiling bucket hash table.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.BuckHashSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("gc_metadata"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: I suggest gc_sys_bytes. (Unit in the metric name, and try to mirror the name in memstats.)

File: prometheus/go_collector.go [hunk 30]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("mcache_sys"),
+					"Number of mcache structures obtained from system.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.MCacheSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("buck_hash_sys"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: buck_hash_sys_bytes

File: prometheus/go_collector.go [hunk 31]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("mcache_inuse"),
+					"Number of mcache structures in use.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.MCacheInuse) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("mcache_sys"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: mcache_sys_bytes

File: prometheus/go_collector.go [hunk 32]
```diff
@@ -9,27 +10,245 @@ import (
...
+					"Number of mcache structures obtained from system.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Number of bytes used for mcache structures obtained from system.

File: prometheus/go_collector.go [hunk 33]
```diff
@@ -9,27 +10,245 @@ import (
 type goCollector struct {
 	goroutines Gauge
 	gcDesc     *Desc
+
+	// memstats object to reuse
```
Comment (beorn7) [inline prometheus/go_collector.go]: Re-using will create a race. Doc for Collector.Collect(): "This method may be called concurrently and must therefore be implemented in a concurrency safe way. "

File: prometheus/go_collector.go (hunk 13 above)
Comment (beorn7) [inline prometheus/go_collector.go]: Perhaps also mention that this is the sum of all the other "sys" metrics.

File: prometheus/go_collector.go [hunk 34]
```diff
@@ -9,27 +10,245 @@ import (
...
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.TotalAlloc) },
+				valType: CounterValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("sys_bytes"),
+					"Number of bytes obtained from system",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.Sys) },
+				valType: CounterValue,
```
Comment (beorn7) [inline prometheus/go_collector.go]: Are you sure this is a counter?

File: prometheus/go_collector.go [hunk 35]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("heap_alloc_bytes"),
+					"Number heap bytes allocated and still in use.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapAlloc) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("heap_sys_bytes"),
+					"Total bytes in heap obtained from system.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: "Number of heap bytes obtained from system." (Only use "total" if it is a counter, and then use it in the metric name, too.)

File: prometheus/go_collector.go [hunk 36]
```diff
@@ -9,27 +10,245 @@ import (
...
+					memstatNamespace("heap_idle_bytes"),
+					"Number bytes in heap waiting to be used.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapIdle) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("heap_inuse_bytes"),
+					"Number of bytes in heap that are in use.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: "Number of heap bytes that are in use." (for consistency)

File: prometheus/go_collector.go [hunk 37]
```diff
@@ -9,27 +10,245 @@ import (
...
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapInuse) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("heap_released_bytes"),
+					"Number of bytes in heap released to OS.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: "Number of heap bytes released to the OS."
If it's a counter (see below) metric should be called "heap_released_bytes_total" and doc string should be "Total number of heap bytes..."

File: prometheus/go_collector.go [hunk 38]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("heap_objects"),
+					"Number of allocated objects.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.HeapObjects) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("stack_bytes_inuse"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: "stack_inuse_bytes"

File: prometheus/go_collector.go [hunk 39]
```diff
@@ -9,27 +10,245 @@ import (
...
+				desc: NewDesc(
+					memstatNamespace("stack_sys_bytes"),
+					"Number of bytes in obtained from system for stack allocator.",
+					nil, nil,
+				),
+				eval:    func(ms *runtime.MemStats) float64 { return float64(ms.StackSys) },
+				valType: GaugeValue,
+			}, {
+				desc: NewDesc(
+					memstatNamespace("mspan_inuse"),
```
Comment (beorn7) [inline prometheus/go_collector.go]: Append "_bytes". Same below for all "_inuse" and "_sys" metrics.

File: prometheus/go_collector.go [hunk 40]
```diff
@@ -9,27 +10,245 @@ import (
...
+					"Number of mspan structures in use.",
```
Comment (beorn7) [inline prometheus/go_collector.go]: Pretty sure this is bytes, not number of structures. Same below.

**Comment by beorn7** [comment]
Sorry for the many comments. Those many MemStats fields are tricky...

File: prometheus/go_collector.go (hunk 11 above)
Comment (KevinPike) [inline prometheus/go_collector.go]: Ah, thanks for the link. I reviewed the style guides

File: prometheus/go_collector.go (hunk 16 above)
Comment (KevinPike) [inline prometheus/go_collector.go]: I think so, I'll update it

File: prometheus/go_collector.go (hunk 34 above)
Comment (KevinPike) [inline prometheus/go_collector.go]: I'm not sure. I will make it a GaugeValue for now. How would you go about figuring this out?

**Comment by KevinPike** [comment]
Thank you for all the comments! These MemStats are tricky. I updated the metric descriptions to match your comments. As far as changing the types of value types, what's the best way to determine the value type?

**Comment by beorn7** [comment]
Yeah, the documentation of the memstats fields is not very... comprehensive... ;)

I see two strategies:
1. Run a reasonably complex binary with your changes and loot how values behave.
2. Check out the code of the Go runtime and see what it is actually doing.

I'll look closer at your recent code updates tomorrow.

File: prometheus/go_collector.go (hunk 16 above)
Comment (beorn7) [inline prometheus/go_collector.go]: I'm pretty sure it's a counter. Let's implement it as such, and change later if we are mistaken.

So my remark above about docstring and metric name applies, too.

**Comment by beorn7** [comment]
Looks good, just adjust `heap_released_bytes` as commented.

**Comment by KevinPike** [comment]
Thanks @beorn7, I updated `heap_released_bytes`

File: prometheus/go_collector.go (hunk 37 above)
Comment (beorn7) [inline prometheus/go_collector.go]: This still applies.

**Comment by beorn7** [comment]
Somehow, GH doesn't display my most recent comment up there. Rephrasing here:

"heap_released_bytes" should be called "heap_released_bytes_total", as it is a counter now.

Docstring should be "Total number of heap bytes released to OS."

**Comment by KevinPike** [comment]
@beorn7 ah, I missed that. Okay, this time it's really updated :smile:

**Comment by beorn7** [comment]
:+1: Thanks for all the efforts.

**Pull request merged**